services/datamanager/builtin/file_deletion.go

```diff
@@ -0,0 +1,135 @@
+	usage := du.NewDiskUsage(captureDirPath)
+	// we get usage this way to ensure we get the amount of remaining space in the partition.
+	// calling usage.Usage() returns the usage of the whole disk, not the user partition
+	usedSpace := 1.0 - float64(usage.Available())/float64(usage.Size())
+	if math.IsNaN(float64(usedSpace)) {
+		return false, nil
+	}
+	if usedSpace < fsThresholdToTriggerDeletion {
+		logger.Warnw("disk not full enough, exiting", "used space", usedSpace)
+		return false, nil
+	}
+	// Walk the dir to get capture stats
+	return exceedsDeletionThreshold(ctx, captureDirPath, float64(usage.Size()), logger)
+}
+
+func exceedsDeletionThreshold(ctx context.Context, captureDirPath string, fsSize float64, logger logging.Logger) (bool, error) {
+	var dirSize int64 = 0
+
+	readSize := func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				return nil
+			}
+			return err
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		if !d.IsDir() {
+			fileInfo, err := d.Info()
+			if err != nil {
+				return err
+			}
+			dirSize += fileInfo.Size()
+			if float64(dirSize)/fsSize > captureDirToFSUsageRatio {
+				logger.Warnw("At threshold to delete, going to delete", "usage ratio", float64(dirSize)/fsSize, "threshold", captureDirToFSUsageRatio)
+				return errAtSizeThreshold
+			}
+		}
+		return nil
+	}
+
+	err := filepath.WalkDir(captureDirPath, readSize)
+	if err != nil && !errors.Is(err, errAtSizeThreshold) {
+		return false, err
+	}
+	if err == nil {
+		logger.Warnw("Not at threshold", "size", float64(dirSize)/fsSize, "threshold", captureDirToFSUsageRatio)
+
+	}
+	return err != nil && errors.Is(err, errAtSizeThreshold), nil
+}
+
+func deleteFiles(ctx context.Context, syncer datasync.Manager, captureDirPath string, logger logging.Logger) (int, error) {
+	index := 0
+	deletedFileCount := 0
+	delete := func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			// this can happen if after we start walking the dir, the file changes from .prog to .capture
+			// which throws a file not found error when we try to get the fileinfo. If we hit this, just
+			// swallow the error and continue walking
+			if errors.Is(err, fs.ErrNotExist) {
+				return nil
+			}
+			return err
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		if !d.IsDir() {
+			fileInfo, err := d.Info()
+			if err != nil {
+				// same reason as above
+				if errors.Is(err, fs.ErrNotExist) {
+					return nil
+				}
+				return err
+			}
+			isFileInProgress := strings.Contains(fileInfo.Name(), datacapture.InProgressFileExt)
+			// if at nth file and the file is not currently being written, mark as in progress if possible
+			if index%n == 0 && !isFileInProgress {
+				if syncer != nil && !syncer.MarkInProgress(path) {
+					logger.Warnw("Tried to mark file as in progress but lock already held", "file", d.Name())
+					index++
```

Comment:
I think we actually don't want to increment the index here, since we will otherwise then need to iterate over n more files before attempting deletion again.

Reply:
done!